### PR TITLE
Cache expansion hash globally

### PR DESCRIPTION
**Pull request by cjgillot**
... instead of computing it multiple times.

Split from #86676 
r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -36,9 +36,10 @@ use source_map::SourceMap;
 pub mod edition;
 use edition::Edition;
 pub mod hygiene;
-pub use hygiene::SyntaxContext;
 use hygiene::Transparency;
-pub use hygiene::{DesugaringKind, ExpnData, ExpnId, ExpnKind, ForLoopLoc, MacroKind};
+pub use hygiene::{
+    DesugaringKind, ExpnData, ExpnHash, ExpnId, ExpnKind, ForLoopLoc, MacroKind, SyntaxContext,
+};
```
Comment (petrochenkov) [inline compiler/rustc_span/src/lib.rs:42]: Nit: could you avoid multi-line imports.

File: compiler/rustc_span/src/hygiene.rs [hunk 2]
```diff
@@ -164,6 +179,8 @@ pub struct HygieneData {
     /// between creation of an expansion ID and obtaining its data (e.g. macros are collected
     /// first and then resolved later), so we use an `Option` here.
     expn_data: Vec<Option<ExpnData>>,
+    expn_hash: Vec<ExpnHash>,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:182]:
```suggestion
    expn_hashes: Vec<ExpnHash>,
```

File: compiler/rustc_span/src/hygiene.rs [hunk 3]
```diff
@@ -164,6 +179,8 @@ pub struct HygieneData {
...
+    expn_of_hash: UnhashMap<ExpnHash, ExpnId>,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:183]:
```suggestion
    expn_hash_to_expn_id: UnhashMap<ExpnHash, ExpnId>,
```

File: compiler/rustc_span/src/hygiene.rs [hunk 4]
```diff
@@ -997,8 +1025,8 @@ impl HygieneEncodeContext {
     pub fn encode<
         T,
         R,
-        F: FnMut(&mut T, u32, &SyntaxContextData) -> Result<(), R>,
-        G: FnMut(&mut T, u32, &ExpnData) -> Result<(), R>,
+        F: FnMut(&mut T, u32, SyntaxContextData) -> Result<(), R>,
+        G: FnMut(&mut T, u32, ExpnData, ExpnHash) -> Result<(), R>,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1029]: Could you convert these to `impl Trait`?

File: compiler/rustc_span/src/hygiene.rs [hunk 5]
```diff
@@ -1058,7 +1086,7 @@ pub struct HygieneDecodeContext {
 pub fn decode_expn_id<
     'a,
     D: Decoder,
-    F: FnOnce(&mut D, u32) -> Result<ExpnData, D::Error>,
+    F: FnOnce(&mut D, u32) -> Result<(ExpnData, ExpnHash), D::Error>,
     G: FnOnce(CrateNum) -> &'a HygieneDecodeContext,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1090]: Could you convert these to `impl Trait`?

File: compiler/rustc_span/src/hygiene.rs [hunk 6]
```diff
@@ -1107,6 +1139,8 @@ pub fn decode_expn_id<
         }
 
         hygiene_data.expn_data.push(Some(expn_data));
+        hygiene_data.expn_hash.push(hash);
+        assert!(hygiene_data.expn_of_hash.insert(hash, expn_id).is_none());
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1143]: Could you avoid placing significant code like `insert` inside asserts?

File: compiler/rustc_span/src/hygiene.rs [hunk 7]
```diff
@@ -1193,28 +1227,34 @@ pub fn decode_syntax_context<
     Ok(new_ctxt)
 }
 
-fn for_all_ctxts_in<E, F: FnMut((u32, SyntaxContext, &SyntaxContextData)) -> Result<(), E>>(
+fn for_all_ctxts_in<E, F: FnMut(u32, SyntaxContext, SyntaxContextData) -> Result<(), E>>(
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1230]: Could you convert this to `impl Trait`?

Comment (petrochenkov) [reply compiler/rustc_span/src/hygiene.rs:1230]: Also, `SyntaxContextData` is relatively heavy, is there any specific reason to pass it by value here?

File: compiler/rustc_span/src/hygiene.rs [hunk 8]
```diff
@@ -1193,28 +1227,34 @@ pub fn decode_syntax_context<
...
     });
     for (ctxt, data) in all_data.into_iter() {
-        f((ctxt.0, ctxt, &data))?;
+        f(ctxt.0, ctxt, data)?;
     }
     Ok(())
 }
 
-fn for_all_expns_in<E, F: FnMut(u32, ExpnId, &ExpnData) -> Result<(), E>>(
+fn for_all_expns_in<E, F: FnMut(u32, ExpnId, ExpnData, ExpnHash) -> Result<(), E>>(
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1243]: Could you convert this to `impl Trait`?

File: compiler/rustc_span/src/hygiene.rs [hunk 9]
```diff
@@ -1306,6 +1346,12 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
     }
 }
 
+fn hash_expn(expn_data: &ExpnData, ctx: &mut impl HashStableContext) -> Fingerprint {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1349]: This can be a method on `ExpnData`?

File: compiler/rustc_span/src/hygiene.rs [hunk 10]
```diff
@@ -1315,105 +1361,72 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
     });
 
-    if modified {
+    if disambiguator != 0 {
         debug!("Set disambiguator for {:?} (hash {:?})", expn_id, first_hash);
-        debug!("expn_data = {:?}", expn_id.expn_data());
+        debug!("expn_data = {:?}", expn_data);
+
+        expn_data.disambiguator = disambiguator;
+        first_hash = hash_expn(&expn_data, &mut ctx);
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1388]: This is not a "first" hash anymore, perhaps other naming should be used.

File: compiler/rustc_span/src/hygiene.rs [hunk 11]
```diff
@@ -1315,105 +1361,72 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
+            );
+        });
+    }
+
+    let first_hash = ExpnHash(first_hash);
+    HygieneData::with(|data| {
+        data.expn_data[expn_id.0 as usize].as_mut().unwrap().disambiguator = disambiguator;
+        debug_assert_eq!(data.expn_hash[expn_id.0 as usize].0, Fingerprint::ZERO);
+        data.expn_hash[expn_id.0 as usize] = first_hash;
+        assert!(data.expn_of_hash.insert(first_hash, expn_id).is_none());
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1406]: Could you avoid placing significant code like `insert` inside asserts?

File: compiler/rustc_span/src/hygiene.rs [hunk 12]
```diff
@@ -1315,105 +1361,72 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
+            let (expn_id, transparency) = self.outer_mark();
+            expn_id.hash_stable(ctx, hasher);
+            transparency.hash_stable(ctx, hasher);
+        }
+    }
+}
+
+impl<CTX: HashStableContext> HashStable<CTX> for ExpnId {
+    fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
+        let hash = if *self == ExpnId::root() { Fingerprint::ZERO } else { self.expn_hash().0 };
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1428]: Is this an optimization?
`self.expn_hash().0` should return `Fingerprint::ZERO` for the root too.

File: compiler/rustc_metadata/src/rmeta/mod.rs [hunk 13]
```diff
@@ -226,6 +227,7 @@ crate struct CrateRoot<'tcx> {
 
     syntax_contexts: SyntaxContextTable,
     expn_data: ExpnDataTable,
+    expn_hash: ExpnHashTable,
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/rmeta/mod.rs:230]:
```suggestion
    expn_hashes: ExpnHashTable,
```

File: compiler/rustc_resolve/src/lib.rs [hunk 14]
```diff
@@ -1149,6 +1149,14 @@ impl ResolverAstLowering for Resolver<'_> {
         self.opt_local_def_id(node).unwrap_or_else(|| panic!("no entry for node id: `{:?}`", node))
     }
 
+    fn def_path_hash(&self, def_id: DefId) -> DefPathHash {
+        if let Some(def_id) = def_id.as_local() {
+            self.definitions.def_path_hash(def_id)
+        } else {
+            self.cstore().def_path_hash(def_id)
+        }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/lib.rs:1157]: Nit: `match` often look more compact than `if let` with a single `else` clause.

**Comment by petrochenkov** [comment]
LGTM, left some comments, mostly about readability/searchability.

File: compiler/rustc_span/src/hygiene.rs [hunk 15]
```diff
@@ -1012,16 +1020,11 @@ pub struct HygieneEncodeContext {
...
-        F: FnMut(&mut T, u32, SyntaxContextData) -> Result<(), R>,
-        G: FnMut(&mut T, u32, ExpnData, ExpnHash) -> Result<(), R>,
-    >(
+    pub fn encode<T, R>(
         &self,
         encoder: &mut T,
-        mut encode_ctxt: F,
-        mut encode_expn: G,
+        mut encode_ctxt: impl FnMut(&mut T, u32, SyntaxContextData) -> Result<(), R>,
+        mut encode_expn: impl FnMut(&mut T, u32, ExpnData, ExpnHash) -> Result<(), R>,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1027]: `ExpnData` is heavy, any specific reason to pass it by value?

File: compiler/rustc_span/src/hygiene.rs [hunk 16]
```diff
@@ -1230,15 +1232,15 @@ fn for_all_ctxts_in<E, F: FnMut(u32, SyntaxContext, SyntaxContextData) -> Result
     Ok(())
 }
 
-fn for_all_expns_in<E, F: FnMut(u32, ExpnId, ExpnData, ExpnHash) -> Result<(), E>>(
+fn for_all_expns_in<E>(
     expns: impl Iterator<Item = ExpnId>,
-    mut f: F,
+    mut f: impl FnMut(u32, ExpnId, ExpnData, ExpnHash) -> Result<(), E>,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1237]: `ExpnData` is heavy, any specific reason to pass it by value?

**Comment by petrochenkov** [comment]
Three remaining questions about passing heavy structures by value - https://github.com/rust-lang/rust/pull/87044#discussion_r667473378, https://github.com/rust-lang/rust/pull/87044#discussion_r667545140, https://github.com/rust-lang/rust/pull/87044#discussion_r667545160.

File: compiler/rustc_span/src/hygiene.rs (hunk 15 above)
Comment (cjgillot) [reply compiler/rustc_span/src/hygiene.rs:1027]: This is required by the encoding code in the on-disk cache, which accepts `&(ExpnData, ExpnHash)`, but not `&(&ExpnData, ExpnHash)`. However, #86676 can pass a reference.

**Comment by petrochenkov** [comment]
r=me with commits starting from https://github.com/rust-lang/rust/pull/87044/commits/c683196980ad8fe4a509d600e081f5e8ab370a55 squashed.

**Comment by cjgillot** [comment]
The perf results are globally green with ~1% improvements, except an isolated regression on `deeply-nested-async`.

**Pull request merged**